### PR TITLE
feat: npm package for publishing templates

### PR DESCRIPTION
**Pull request by jerryzhang222**
This package provides script to validate and publish templates. 

To validate:
```yaml
shared:
    image: node:6
jobs:
    main:  
        steps:
            - install: npm install screwdriver-template-main
            - validate: ./node_modules/.bin/template-validate
        environment:
            SD_TEMPLATE_PATH: ./path/to/template.yaml
```

To publish: 
```yaml
shared:
    image: node:6
jobs:
    main:  
        steps:
            - install: npm install screwdriver-template-main
            # this is run by the PRs as well
            - validate: ./node_modules/.bin/template-validate
    publish:
        steps:
            - install: npm install screwdriver-template-main
            # publishing also validates
            - publish: ./node_modules/.bin/template-publish
```

Blocked by: https://github.com/screwdriver-cd/screwdriver/pull/520
Related: https://github.com/screwdriver-cd/screwdriver/issues/470

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,7 +37,12 @@
     "eslint-config-screwdriver": "^2.0.9",
     "jenkins-mocha": "^3.0.0"
   },
-  "dependencies": {},
+  "dependencies": {
+    "js-yaml": "^3.8.3",
+    "mockery": "^2.0.0",
```
Comment (FenrirUnbound) [inline package.json]: test dependencies should go into the `devDependencies`

File: publish.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+'use strict';
```
Comment (FenrirUnbound) [inline publish.js]: Are you planning on making this executable? If so, you're missing a couple of things:

* need the `#!/usr/bin/env node` preamble at the top of this
* Add it to the `bin` key in the `package.json`, so it's like:

```
{
  "name": "screwdriver-template-main",
  ...
  "bin": "./package.js"
}
```

Even better, you could specify the binary name with: 

```
{
  "name": "screwdriver-template-main",
  ...
  "bin": {
    "template-publisher": "./package.js"
  }
}
```

Doc for reference: https://docs.npmjs.com/files/package.json#bin

File: README.md [hunk 3]
```diff
@@ -9,6 +9,34 @@
...
+
+jobs:
+    main:
+        steps:
+            - random: echo 'random step'
+    publishfirst:
+        template: screwdriver/publishTemplate
+        environment:
+            SD_TEMPLATE_PATH: ./path/to/my-template.yaml
+    publishsecond:
```
Comment (petey) [inline README.md]: This feels really cumbersome and time consuming for the template maintainer.

Lets have two environment variables:
`SD_TEMPLATE_FILE`: Override default behavior to publish just this one file
`SD_TEMPLATE_DIR`: Publish all yaml files in that directory

I suggest that the `SD_TEMPLATE_DIR` option be the default, with it publishing anything in `templates`.

Comment (petey) [inline README.md]: Also, we need to make sure that the publish endpoint validates the template. We don't want the user to be able to upload invalid templates.

**Comment by petey** [comment]
Since my comments got swallowed:

This feels really cumbersome and time consuming for the template maintainer.

Lets have two environment variables:
SD_TEMPLATE_FILE: Override default behavior to publish just this one file
SD_TEMPLATE_DIR: Publish all yaml files in that directory

I suggest that the SD_TEMPLATE_DIR option be the default, with it publishing anything in templates.

Also, we need to make sure that the publish endpoint validates the template. We don't want the user to be able to upload invalid templates.

File: index.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const path = process.env.SD_TEMPLATE_PATH || './sd-template.yaml';
+    const yaml = Yaml.safeLoad(fs.readFileSync(path, 'utf8'));
+    const params = {
+        body: yaml,
+        json: true,
+        auth: {
+            bearer: process.env.SD_TOKEN
+        },
+        method: 'POST',
+        url: 'https://api.screwdriver.cd:443/v4/templates'
```
Comment (FenrirUnbound) [inline index.js]: Do we need to specify the port? 443 is the default port for HTTPS

File: index.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+            throw new Error(`Error sending request: ${err}`);
+        } else if (response.statusCode !== 201) {
+            throw new Error('Template was not published. ' +
+                `${body.statusCode} (${body.error}): ${body.message}`);
+        } else {
+            console.log('Template successfully published.');
+        }
+    });
+
+    return null;
```
Comment (FenrirUnbound) [inline index.js]: Could also just return the `request(..)`

File: publish.js (hunk 2 above)
Comment (FenrirUnbound) [inline publish.js]: Looks like it's addressed 👍

File: test/index.test.js [hunk 6]
```diff
@@ -1,9 +1,94 @@
...
+    });
+
+    after(() => {
+        mockery.disable();
+    });
+
+    it('throws error when request yields an error', () => {
+        const error = new Error('error');
+
+        requestMock.yields(error);
```
Comment (FenrirUnbound) [inline test/index.test.js]: Should be `.yieldsAsync`

File: test/index.test.js [hunk 7]
```diff
@@ -1,9 +1,94 @@
...
+        assert.throws(index.publishTemplate, Error, 'Error sending request: Error: error');
```
Comment (FenrirUnbound) [inline test/index.test.js]: It's an asynchronous operation, so any errors it throws won't be catch-able. If you switch the mock over to `.yieldsAsync`, this error will show up.

**Comment by tkyi** [comment]
Yeah I agree with @petey . Ideally, the user could just specify `SD_TEMPLATE_FILE` or `SD_TEMPLATE_DIR`. However, I'd probably assume if none were provided, default path would be `./sd-template.yaml`. If we do go with `SD_TEMPLATE_DIR` as the default option, I think it should publish anything in `./sd-templates`.

👍 to Petey's idea that the publish template should also validate. So that means we don't need a separate template for validating.

**Comment by d2lam** [comment]
@petey Yes, the validator endpoint was a WIP so we decided to split up the work. Then add the validator after this PR is merged. 
@tkyi We would still need to separate a template for validating. The reason is PR jobs would run `main`. So we don't want to have publishing in there. We need a template for just validating, so that PRs can just run the validating step without publishing. 

We will modify it so that it can publish all the templates from a directory.

**Comment by tkyi** [comment]
@d2lam so the responsibility would fall on the user to make sure they call the `validate` template before publishing?

File: index.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+const Yaml = require('js-yaml');
+
+/**
+ * Publishes the template yaml by posting to the SDAPI /templates endpoint
+ * @method publishTemplate
+ * @return {null}
+ */
+function publishTemplate() {
+    const path = process.env.SD_TEMPLATE_PATH || './sd-template.yaml';
+    const yaml = Yaml.safeLoad(fs.readFileSync(path, 'utf8'));
```
Comment (minz1027) [inline index.js]: `safeLoad` will throw err if the template.yaml is not a valid yaml. Shall we have  some error handling for that? 
https://github.com/nodeca/js-yaml

**Comment by d2lam** [comment]
@tkyi  If you meant that as a question: No, I'm saying we should have a template-validate that does pure validating. Our template-publish should do both: validate, then publish. 
And in their `screwdriver.yaml`
```
main: 
   template: template-validate
publish:
   template: template-publish
```

File: index.js (hunk 8 above)
Comment (d2lam) [inline index.js]: It should be handled by the validator step, which is not plugged in yet.

**Comment by tkyi** [comment]
@d2lam Haha yes, it was a question.
Oh okay! I see. That makes sense. 👍 
So the validate template needs to be created before it can be used in the publish template.

File: README.md [hunk 9]
```diff
@@ -9,6 +9,25 @@
 npm install screwdriver-template-main
 ```
 
+### Publishing a template using Screwdriver
+
+To publish a new template, installs the `screwdriver-template-main` npm package, and run the `template-publish` script. By default, the path `./sd-template.yaml` will be read. However, a user can specify a custom path using the env variable: `SD_TEMPLATE_PATH`.
+
+To publish multiple templates in the same repo, a template maintainer would use the following pattern in their `screwdriver.yaml`:
```
Comment (tkyi) [inline README.md]: Should we update the README to reflect @petey 's comment?

**Comment by d2lam** [comment]
@petey Actually I have some concern about publishing all the templates in `SD_TEMPLATE_DIR`. 
If we implement it that way, it's unclear if some of the templates fail to publish. For example: if there are `template1`, `template2`, `template3` are under `SD_TEMPLATE_DIR`, `template1` POST call was successful, but `template2` was not. The job failed, but it's unclear that `template1` was actually published.

If we split them up, each job is meant to publish 1 template, then it's more clear. If the job fails, that means that template was not published. And that would also make more sense when we have parallel implemented. They can publish multiple templates in parallel instead of in series.

**Comment by petey** [comment]
@d2lam So is your concern that we don't have an all-or-nothing transaction based publish from this template? We can clearly log what has and what has not been published.

If we have to publish everything individually, they all need to go through the various bookends multiple times. This is time consuming. It also means that the user has to set up and maintain jobs for every individual template to publish.

**Comment by d2lam** [comment]
@petey 
> So is your concern that we don't have an all-or-nothing transaction based publish from this template? We can clearly log what has and what has not been published.

Yes, that is exactly my concern. They would need to dig into the build log to see that. As a user, I would expect if the job fails, none of my templates were published, which is not the case here.

I agree with you that it would need to go through multiple bookends to get their build running. That is annoying. If we could find a way to resolve the above issue then that would be best.

**Comment by petey** [comment]
ok. we can figure out how to do multiple templates later, but I'm still unwilling to merge this until the API itself is validating template contents before it allows that template to be created.

File: index.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+const request = require('request-promise');
```
Comment (petey) [inline index.js]: Please use `request-promise-native` instead.

File: index.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+            const body = response.body;
+
+            if (response.statusCode !== 201) {
+                throw new Error('Template was not published. ' +
+                `${body.statusCode} (${body.error}): ${body.message}`);
+            }
+
+            return Promise.resolve('Template successfully published.');
+        })
+        .catch(err => Promise.reject(err.message));
```
Comment (petey) [inline index.js]: When you throw in a promise chain, it automatically rejects with the thrown error. 

This line causes the error object to transform into a string, so you lose nice things like stack traces. It is better to pass around the error object in a rejection, so this entire line is not necessary.

**Comment by d2lam** [comment]
@petey  Added validation before publishing & also refactored some logic. Could you please re-review! Thanksss :)

File: index.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+        auth: {
+            bearer: 'does not matter'
+        },
+        json: true,
+        body: {
+            yaml: JSON.stringify(config)
+        }
+    }).then((response) => {
+        if (response.errors.length > 0) {
+            throw new Error('Template has invalid format');
```
Comment (tkyi) [inline index.js]: Should we also return the response body here so the user can see what errors there are?

Comment (d2lam) [inline index.js]: @tkyi  Added

File: README.md [hunk 13]
```diff
@@ -9,6 +9,46 @@
 npm install screwdriver-template-main
 ```
 
+### Validating a template using Screwdriver
+
+Run the `template-publish` script. By default, the path `./sd-template.yaml` will be read. However, a user can specify a custom path using the env variable: `SD_TEMPLATE_PATH`.
```
Comment (tkyi) [inline README.md]: Run the `template-validate` script.*

File: index.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+ * @method validateTemplate
+ * @param   {Object}        config          Template config
+ * @return {Promise}         Resolves when template is valid/rejects when error or invalid
+ */
+function validateTemplate(config) {
+    const hostname = process.env.SD_API_URL || 'https://api.screwdriver.cd/v4/';
+
+    return request({
+        method: 'POST',
+        url: `${hostname}validator/template`,
```
Comment (petey) [inline index.js]: Should probably use `path.join` or `url.resolve` here.

File: index.js [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+ * @method publishTemplate
+ * @param   {Object}        config          Template config
+ * @return {Promise}        Resolves if publish successfully
+ */
+function publishTemplate(config) {
+    const hostname = process.env.SD_API_URL || 'https://api.screwdriver.cd/v4/';
+
+    return request({
+        method: 'POST',
+        url: `${hostname}templates`,
```
Comment (petey) [inline index.js]: Should probably use `path.join` or `url.resolve` here.

File: validate.js [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+'use strict';
+
+const index = require('./index');
+const path = process.env.SD_TEMPLATE_PATH || './sd-template.yaml';
+
+return index.loadYaml(path)
+    .then(config => index.validateTemplate(config))
+    .then(console.log)
+    .catch((err) => {
+        console.log(err);
```
Comment (petey) [inline validate.js]: `console.error(err)`

File: publish.js [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+
+const index = require('./index');
+const path = process.env.SD_TEMPLATE_PATH || './sd-template.yaml';
+
+return index.loadYaml(path)
+    .then(config => index.publishTemplate(config))
+    .then(console.log)
+    .catch((err) => {
+        console.log(err);
```
Comment (petey) [inline publish.js]: `console.error(err)`

File: index.js [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    const url = URL.resolve(hostname, 'templates');
+
+    return request({
+        method: 'POST',
+        url,
+        auth: {
+            bearer: process.env.SD_TOKEN
+        },
+        json: true,
+        body: JSON.stringify(config),
```
Comment (d2lam) [inline index.js]: Need 
```
body:
    yaml: JSON.stringify(config)
```
here actually

**Comment by d2lam** [comment]
@petey  addressed the comments

**Comment by d2lam** [comment]
@petey We parses the `template.yaml` and turns back to string because of we wanted the `/validator` and the `/create` endpoint to be the same. Do you remember from this PR: https://github.com/screwdriver-cd/screwdriver/pull/520? Any ways so that we don't have to do that?

**Comment by petey** [comment]
The only reason to parse the yaml ahead of time is to catch that the yaml is invalid before making a server call. You could just send the raw content of the yaml, and not send it through `js-yaml` at all. If we were ever to change to a different yaml parser, however unlikely, that implemented some feature we wanted to rely on, we have to go back and change this, too.

Since JSON is a subset of yaml, it shouldn't really make a big difference in this case, but the formatting of the content that is being sent to the validator and publish api from this code is different from the formatting of the original file.

**Pull request merged**